src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/models/Transition.kt

```diff
@@ -0,0 +1,137 @@
+import org.elasticsearch.common.xcontent.ToXContent
+import org.elasticsearch.common.xcontent.ToXContentObject
+import org.elasticsearch.common.xcontent.XContentBuilder
+import org.elasticsearch.common.xcontent.XContentParser
+import org.elasticsearch.common.xcontent.XContentParser.Token
+import org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken
+import java.io.IOException
+
+data class Transition(
+    val stateName: String,
+    val conditions: Conditions?
+) : ToXContentObject {
+
+    override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        builder
+            .startObject()
+                .field(STATE_FIELD, stateName)
+                .field(CONDITIONS_FIELD, conditions)
+            .endObject()
+        return builder
+    }
+
+    companion object {
+        const val STATE_FIELD = "state_name"
+        const val CONDITIONS_FIELD = "conditions"
+
+        @JvmStatic
+        @Throws(IOException::class)
+        fun parse(xcp: XContentParser): Transition {
+            lateinit var name: String
+            var conditions: Conditions? = null
+
+            ensureExpectedToken(Token.START_OBJECT, xcp.currentToken(), xcp::getTokenLocation)
+            while (xcp.nextToken() != Token.END_OBJECT) {
+                val fieldName = xcp.currentName()
+                xcp.nextToken()
+
+                when (fieldName) {
+                    STATE_FIELD -> name = xcp.text()
+                    CONDITIONS_FIELD -> conditions = Conditions.parse(xcp)
+                    else -> throw IllegalArgumentException("Invalid field: [$fieldName] found in Transition.")
+                }
+            }
+
+            return Transition(
+                stateName = requireNotNull(name) { "Transition state name is null" },
+                conditions = conditions
+            )
+        }
+    }
+}
+
+data class Conditions(
+    val indexAge: String?,
+    val docCount: Long?,
+    val size: String?,
+    val cron: CronSchedule?
+) : ToXContentObject {
+
+    init {
+        val conditionsList = listOf(indexAge, docCount, size, cron)
+        require(conditionsList.filterNotNull().size < 2) { "Cannot provide more than one Transition condition" }
+
+        // Validate indexAge condition
+        try {
+            TimeValue.parseTimeValue(indexAge, "")
+        } catch (e: IllegalArgumentException) {
+            throw IllegalArgumentException("Must have a valid index age Transition condition")
+        }
+
+        // Validate size condition
+        requireNotNull(size?.toIntOrNull()) { "Must have a valid size Transition condition" }
+    }
+
+    override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        builder
+            .startObject()
+                .field(INDEX_AGE_FIELD, indexAge)
+                .field(DOC_COUNT_FIELD, docCount)
+                .field(SIZE_FIELD, size)
+                .field(CRON_FIELD, cron)
```

Comment:
we should skip the field if the value is null. Unless we want to intentionally write null fields.

Reply:
Done.